cedar-policy-validator/src/typecheck.rs

```diff
@@ -1961,83 +1964,126 @@ impl<'a> Typechecker<'a> {
-                    } else {
-                        // This should only happen when doing partial validation
-                        // since we never construct the undeclared action
-                        // request environment otherwise.
-                        TypecheckAnswer::fail(in_expr)
-                    }
-                }
-                Some(EntityType::Specified(var_name)) => {
-                    let all_rhs_known = rhs
-                        .iter()
-                        .all(|e| self.schema.euid_has_known_entity_type(e));
-                    if self.schema.is_known_entity_type(var_name) && all_rhs_known {
-                        let descendants = self.schema.get_entity_types_in_set(rhs.iter());
-                        Typechecker::entity_in_descendants(
-                            var_name,
-                            descendants,
-                            in_expr,
-                            lhs_expr,
-                            rhs_expr,
-                        )
+                        // In partial schema mode, undeclared entity types are
+                        // expected.
+                        TypecheckAnswer::success(annotated_expr)
                     } else {
-                        let annotated_expr =
-                            ExprBuilder::with_data(Some(Type::primitive_boolean()))
-                                .with_same_source_loc(in_expr)
-                                .is_in(lhs_expr, rhs_expr);
-                        if self.mode.is_partial() {
-                            // In partial schema mode, undeclared entity types are
-                            // expected.
-                            TypecheckAnswer::success(annotated_expr)
-                        } else {
-                            TypecheckAnswer::fail(annotated_expr)
-                        }
+                        TypecheckAnswer::fail(annotated_expr)
                     }
                 }
-                // Unspecified entities will be detected by a different part of the validator.
-                // Still return `TypecheckFail` so that typechecking is not considered successful.
-                Some(EntityType::Unspecified) => TypecheckAnswer::fail(
+            }
+            (Some(EntityType::Specified(_)), None) => {
+                // One or more of the elements on the right is not an entity
+                // literal. The `in` is still valid, so typechecking succeeds
+                // with type Boolean.
+                TypecheckAnswer::success(
                     ExprBuilder::with_data(Some(Type::primitive_boolean()))
                         .with_same_source_loc(in_expr)
                         .is_in(lhs_expr, rhs_expr),
-                ),
+                )
+            }
+            (Some(EntityType::Unspecified), Some(rhs)) => {
+                // It's perfectly valid for `principal` or `resource` to be `EntityType::Unspecified`
```

Comment:
This case is a little weird to me. It's changing behavior from returning an error to returning Bool or False. It seems like your edit is the correct behavior, but why wasn't this discrepancy picked up by DRT? Is there some other part of the code that was handling unspecified variables, as suggested by the original (possibly incorrect) comment?

Reply:
It turns out that returning the `TypecheckFail` results in `typecheck_policy()` returning `false` instead of `true`, but that return value is discarded by its caller ([here](https://github.com/cedar-policy/cedar/blob/1ce81ec45f95723ba838f0a125070fe7de7f124e/cedar-policy-validator/src/lib.rs#L151)). Since no actual type errors are added to the set of type errors, validation is considered a success despite the `TypecheckFail` which is ignored.

So, my conclusion is that changing this Fail(Bool) to Success(Bool) does not affect anything, and in fact would better represent what is actually happening. Returning a Fail without adding a type error is functionally equivalent to returning a Success, at least in this instance.